### PR TITLE
Quantum extension

### LINKED ISSUE #365: Implement Quantum Extension
And migrate most of the existing `LeafOp` operations into it.

### PR DESCRIPTION
**Pull request by ss2165**
Closes #365
remove quantum ops that are not used in tests

more can be added later as required

Early commits also remove XOR and replace use with logic extension

Required some fixups to enable use of `ExtensionOp`

### CODE REVIEW

File: src/hugr/serialize.rs [hunk 1]
```diff
@@ -430,12 +431,12 @@ pub mod test {
 
     #[test]
     fn dfg_roundtrip() -> Result<(), Box<dyn std::error::Error>> {
-        let tp: Vec<Type> = vec![NAT; 2];
+        let tp: Vec<Type> = vec![BOOL_T; 2];
         let mut dfg = DFGBuilder::new(AbstractSignature::new_df(tp.clone(), tp))?;
         let mut params: [_; 2] = dfg.input_wires_arr();
         for p in params.iter_mut() {
             *p = dfg
-                .add_dataflow_op(LeafOp::Xor, [*p, *p])
```
Comment (ss2165) [inline src/hugr/serialize.rs:438]: `ExtensionOp` serializes as an `OpaqueOp` so the roundtrip test doesn't work with `and_op()`

**Review by cqc-alec — COMMENTED** [review]
Looks good! My main comment is that I think `Measure` should return an `int<1>` or `BOOL_T` rather than `USIZE_T`.

File: src/std_extensions/quantum.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    use super::EXTENSION;
+
+    fn get_gate(gate_name: &str) -> LeafOp {
+        EXTENSION
+            .instantiate_extension_op(gate_name, [])
+            .unwrap()
+            .into()
+    }
+
+    pub(crate) fn h_gate() -> LeafOp {
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/std_extensions/quantum.rs:80]: I'm not fussed, I find "gate" useful as a shorthand for quantum

File: src/extension/op_def.rs [hunk 3]
```diff
@@ -224,7 +224,8 @@ impl OpDef {
             }
             SignatureFunc::CustomFunc(bf) => bf.compute_signature(&self.name, args, &self.misc)?,
         };
-        assert!(res.contains(self.extension()));
+        // TODO bring this assert back once resource inference is done?
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension/op_def.rs:227]: not yet, have asked @croyzor to make one

File: src/std_extensions/quantum.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    extension
+        .add_op_custom_sig_simple(
+            SmolStr::new_inline("Measure"),
+            "Measure a qubit, returning the qubit and the measurement result.".into(),
+            vec![],
+            |_arg_values: &[TypeArg]| {
+                Ok((
+                    type_row![QB_T],
+                    type_row![QB_T, USIZE_T],
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/std_extensions/quantum.rs]: went for BOOL

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**